### ISSUE TITLE
Ferret writes "notes" to XML header file in some cases causing XML parsing to break

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/LAS#779]
**Reported by @noaaroland on 16 Mar 2010 15:16 UTC**
For example:

Adding a ferret script like this:

```
use "http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd"
```

into a directory scanned by F-TDS (conf/server/data for example) will create a new F-TDS data set which uses the mentioned OPeNDAP URL.

However, when the IOSP runs the header script the standard out contains some "Notes" which make the XML invalid.

```

 *** NOTE: regarding http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd ...
 *** NOTE: UNCLASSIFIED
 *** NOTE: regarding http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd ...
 *** NOTE: UNCLASSIFIED
 *** NOTE: regarding http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd ...
 *** NOTE: UNCLASSIFIED
 *** NOTE: regarding http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd ...
 *** NOTE: UNCLASSIFIED
 *** NOTE: regarding http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd ...
 *** NOTE: UNCLASSIFIED
<?xml version="1.0" encoding="ISO-8859-1" standalone="yes"?>
<data>
<datasets>
<dataset name="http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexic
o_Aggregation_best.ncd" default="true">
<title> </title>
```

The current IOSP implementation runs a ferret command similar to this one:

```
/usr/local/ferret/bin/ferret, -memsize, 16, -gif, -server, -script, header.jnl /home/rhs/workspace/baker/conf/server/temp/21ADA05EB3B6E309A5F7CEB30BF94A21/ferret_operation_1268751679850.jnl
```

in which the header.jnl file does a "go $1" to open the named data set and writes the output from several command to standard output.

I have modified the IOSP code to invoke the header command with an additional argument:

```
/usr/local/ferret/bin/ferret, -memsize, 16, -gif, -server, -script, header.jnl /home/rhs/workspace/baker/conf/server/temp/21ADA05EB3B6E309A5F7CEB30BF94A21/ferret_operation_1268751679850.jnl /home/rhs/workspace/baker/conf/server/temp/21ADA05EB3B6E309A5F7CEB30BF94A21/header.xml
```

Now all that remains is to modify the header.jnl script to use the second argument as the file in which it accumulates the output from the commands that produce the header XML.

Migrated-From: http://dunkel.pmel.noaa.gov/trac/las/ticket/773

**Comment by karlmsmith** [comment]
**Comment by @AnsleyManke on 16 Mar 2010 16:17 UTC**
OR, please try the Ferret executable that's currently installed on porter:

ferret_v6425

This executable writes all errors, warnings, and notes to std error (unless I have missed anything that may come from the old pplus code).

If we solve this with a new Ferret executable, I'll need to make a Ferret release before the V7.1 release and that then would be a requirement for the LAS installation.

**Comment by karlmsmith** [comment]
**Comment by @noaaroland on 16 Mar 2010 17:12 UTC**
Getting the fix right into Ferret is very much fine with me, but when I tried this particular data set using ferret_v6425 everything gets written to standard out.

E.g. the following bourne shell command line invocation:

```
sh-3.00$ ferret_v6425 -script header.jnl 1>out 2>err
```

the err file is empty and the out file contains the notes and the XML where header.jnl is a slightly doctored file to use the data set directly:

```
cancel mode verify
use "http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd"
say <?xml version="1.0" encoding="ISO-8859-1" standalone="yes"?> 
say <data>
show var/xml
show data/var/xml
show axis/all/xml
say </data>
```

**Comment by karlmsmith** [comment]
**Comment by @AnsleyManke on 16 Mar 2010 17:35 UTC**
The change will have to be in the Ferret executable regardless. As Ferret is now, nothing can be done to header.jnl to write the xml output to a named file.

When we first implemented xml output in Ferret, we added it to the SHOW command which always writes to std output. In hindsight we should have added it as an option on the LIST command which can easily send its output to a separate file.

So I have installed the correct version of Ferret on porter, v6426, and running a script that contains

```
  ! agg.jnl
  use "http://porter.pmel.noaa.gov:8780/thredds/dodsC/NCOM_Gulf_of_Mexico_Aggregation_best.ncd"
  show data
```


via 

(ferret_v6426 < agg.jnl > aggout.jnl) >& aggerr.jnl

sends the output of the SHOW command to std output, and the NOTES to std error.

I think this'll fix things for now, and for flexibility work on adding the right code to Ferret to let us send the xml output to a file.

**Comment by karlmsmith** [comment]
**Comment by @noaaroland on 16 Mar 2010 18:58 UTC**
The bug is fixed.  And the code is in place in the Java IOSP to deal with writing the header to a file when it's implemented.  I will leave this open until that task is completed or we can close this if you want to open a new ticket with adding the XML output to the LIST command as a task.

**Comment by karlmsmith** [comment]
**Comment by @AnsleyManke on 18 Mar 2010 23:06 UTC**
The Ferret executable 

 /home/porter/tmap/ferret/x86_64-linux/bin/ferret_v65

now has the machinery to write the XML output to a file.  Here is the header.jnl that this executable can use, where the first argument is the script that opens the dataset, and the second is the name of the xml file.

This script is also in 
/home/porter/ansley/rhs/header.jnl

```
!header.jnl
! argument 1 is a script that opens the dataset
! argument 2 is the name of the xml file to write
! Requires Ferret v6.5 or higher

cancel mode verify
IF `($ferret_version) LT 6.5` THEN
  say header.jnl requires Ferret v6.5 or higher
  exit/script
ENDIF

go "$1" 
DEFINE SYMBOL output_xml_file = $2

say/quiet/outfile=($output_xml_file)/clobber <?xml version="1.0" encoding="ISO-8859-1" standalone="yes"?>
say/quiet/outfile=($output_xml_file)/append <data>

show var/xml/append/outfile=($output_xml_file)
show data/var/xml/append/outfile=($output_xml_file)
show axis/all/xml/append/outfile=($output_xml_file)
say/quiet/append </data>

```

Roland, I'm changing ownership of this to you, but hand it back if needed.

**Comment by karlmsmith** [comment]
**Modified by @kevin-obrien on 9 Nov 2010 22:39 UTC**

**Comment by karlmsmith** [comment]
**Comment by @AnsleyManke on 9 Nov 2010 23:33 UTC**
This was fixed with Ferret v6.5 which is relesed.
